### PR TITLE
[Closes #6047] Implements custom `compute_decomposition` method in `Hermitian`

### LINKED ISSUE #6047: Implement `Hermitian.decomposition`
### Important Note

⚠️ This issue is part of an internal assignment and not meant for external contributors.

### Context

Currently, the `compute_decomposition` method of `Hermitian` is inherited from the `Operator` base class, which returns a `qml.DecompositionUndefinedError`. This means if we try to decompose it, we get:

```
>>> op = qml.X(0) + qml.Y(1) + 2 * qml.X(0) @ qml.Z(3)
>>> hermitian_op = qml.Hermitian(qml.matrix(op), wires=['a', 'b', 'aux'])
>>> hermitian_op.decomposition()
DecompositionUndefinedError
````

We want to implement a decomposition method for `Hermitian`, so that the above example instead gives a sum of Pauli operators, along the lines of:

```
>>> hermitian_op.decomposition()
Y('b') + X('a') + 2.0 * (X('a') @ Z('aux'))
```

### Implementation

The task would involve two main steps:

1) Use the existing `pauli_decompose` method to implement a custom `compute_decomposition` method on `Hermitian`. Note the relationship between `decomposition` and `compute_decomposition` as defined in the abstract base class `Operator`.
2) The `pauli_decompose` method is inefficient for larger matrices. Test performance as a function of matrix size and add a check for a reasonable limit. Above the limit we should raise a warning indicating that the decomposition may be inefficient.

In addition to these code changes:

* A clear example should be included in the `Hermitian.compute_decomposition` method
* Relevant tests should be added to [test_observables.py](https://github.com/PennyLaneAI/pennylane/blob/master/tests/ops/qubit/test_observables.py)
* A Changelog entry should be added to [the changelog](https://github.com/PennyLaneAI/pennylane/blob/master/doc/releases/changelog-dev.md) under "Improvements"

### Relevant Files:

* [`pennylane/operation.py`]( https://github.com/PennyLaneAI/pennylane/blob/master/pennylane/operation.py). Contains the base `Operation` class, for context.

* [`pennylane/pauli/conversion.py`](https://github.com/PennyLaneAI/pennylane/blob/master/pauli/conversion.py). Contains the `pauli_decompose` function for decomposing a matrix into a sum of Pauli words. 

* [`pennylane/ops/qubit/observables.py`](https://github.com/PennyLaneAI/pennylane/blob/master/pennylane/ops/qubit/observables.py) contains the `Hermitian` class

### PR DESCRIPTION
**Pull request by andrijapau**
**Context:**

Prior to this fix, the `compute_decomposition` method of `Hermitian` returned a `qml.DecompositionUndefinedError` (due to its `Operator` base class inheritance). The desired fix, as laid out by the issue, was to use the existing `pauli_decompose` method to perform the desired decomposition. 

**Description of the Change:**

Implemented a custom `compute_decomposition` static method  in `Hermitian` to have it call the existing `pauli_decompose` method. 

<ins>Examples</ins>

```python
>>> op = qml.X(0) + qml.Y(1) + 2 * qml.X(0) @ qml.Z(3)
>>> op_matrix = qml.matrix(op)
>>> qml.Hermitian.compute_decomposition(op_matrix, wires=['a', 'b', 'aux'])
[(
      1.0 * (I('a') @ Y('b') @ I('aux'))
    + 1.0 * (X('a') @ I('b') @ I('aux'))
    + 2.0 * (X('a') @ I('b') @ Z('aux'))
)]

>>> op = np.array([[1, 1], [1, -1]]) / np.sqrt(2)
>>> qml.Hermitian.compute_decomposition(op, wires=0)
[(
      0.7071067811865475 * X(0)
    + 0.7071067811865475 * Z(0)
)]
```

 ⏰ *Benchmarking Performance* ⏰

Performance was benchmarked as a function of matrix size (see figure below). I heuristically chose to raise an inefficiency warning if the decomposition time would take longer than a second (open to suggestions 🙂). Any observable with more than seven wires will raise a user warning indicating that the decomposition may be inefficient. 

I've also plotted the theoretical time complexity of the `pauli_decompose` method and the data is in agreement.

benchmark.pdf

**Benefits:**

Can now compute the decomposition of any Hermitian matrix.

**Possible Drawbacks:**

N/A

**Related GitHub Issues:**

Closes #6047

### CODE REVIEW

**Comment by andrijapau** [comment]
Hi @lillian542,

I've opened up this PR as a draft in case you have some early comments. 

Two things I could use some clarification on: 

(1) Any feedback on my approach to benchmarking the method's performance? To me over a second of computation time could warrant some warning from the program. Although, perhaps my experience is a bit biased since I'm used to relatively small quantum system sizes 😁.

(2) Could you confirm the desired output type of `qml.Hermitian.compute_decomposition`? From the example in #6047 it seems like `PauliSentence` would be ideal, but that goes against the expected base class return type of `list[Operator]` causing the `jax-tests` above to fail. I have a fix ready to go which passes the test suite locally but it essentially acknowledges and waives the error in the relevant `ops/functions/conftest.py` file.

Just wanted to double check, thanks!

**Review by mudit2812 — CHANGES_REQUESTED** [review]
Thanks for opening this PR @andrijapau ! It looks really polished. I've left a few comments that need to be addressed before I can approve the changes, but overall, looking great 😄

File: pennylane/ops/qubit/observables.py
Comment (andrijapau) [reply pennylane/ops/qubit/observables.py:209]: Thanks for the clarification, but it looks like your sentence might have been accidentally cut off at the end 🙂. 

> As you spotted, the expected output of the method is `list[Operator]`. `PauliSentence`s are still a somewhat private attribute in PennyLane that are sometimes used as they can be much more efficient, but for the purposes of this PR, the output of this method should adhere to the specified return type of `Operator.compute_decomposition`.
> 
> When `pauli=False`, `pauli_decompose` returns a `Hamiltonian`, which we also don't want. >**_Instead, the_**<

Comment (mudit2812) [reply pennylane/ops/qubit/observables.py:209]: Thanks, it was left over from an outdated comment, please ignore it 😄

File: pennylane/ops/qubit/observables.py
Comment (trbromley) [inline pennylane/ops/qubit/observables.py:167]:
```suggestion
        r"""Decomposes an observable as a sum of Pauli operators.
```

File: pennylane/ops/qubit/observables.py
Comment (trbromley) [inline pennylane/ops/qubit/observables.py:180]: Is this repr correct? I think it should be `Y("b")` - we've tried to make it copy-pasteable (assuming you do `from pennylane import X, Y, Z, I`).

File: pennylane/ops/qubit/observables.py
Comment (trbromley) [inline pennylane/ops/qubit/observables.py:191]: Do we need all caps for these variables? I believe that's [recommended for module-level constants](https://peps.python.org/pep-0008/#constants).

File: pennylane/ops/qubit/observables.py
Comment (albi3ro) [reply pennylane/ops/qubit/observables.py:180]: If how a pauli sentence is rendered, but the operator is `Y('b')`. So once we adjust the return to be a list with an operator in it, the repr will be `
Y('b') + X('a') + (2 * X('a')) @ Z('aux')`.

File: pennylane/ops/qubit/observables.py
Comment (mudit2812) [inline pennylane/ops/qubit/observables.py:209]: 🎉

File: tests/ops/qubit/test_observables.py
Comment (mudit2812) [inline tests/ops/qubit/test_observables.py:185]:
```suggestion
class TestHermitian:  # pylint: disable=too-many-public-methods
```

File: tests/ops/qubit/test_observables.py
Comment (mudit2812) [inline tests/ops/qubit/test_observables.py:336]: For the sake of simplicity, it could be reasonable to just use an identity matrix here with dimension `(2**8, 2**8)` as the intention of the test is just to check that the warning gets raised rather than verifying the correctness of the decomposition.
```suggestion
        observable = qml.Identity(wires=list(range(8)))
```

File: tests/ops/qubit/test_observables.py
Comment (mudit2812) [inline tests/ops/qubit/test_observables.py:367]: These can be collected into a single test. Additionally, I think it's a good idea to stress test the decomposition a bit more by parametrizing over matrices with higher dimensions, say 1-4 wires.

File: tests/ops/qubit/test_observables.py
Comment (mudit2812) [inline tests/ops/qubit/test_observables.py:352]: Nice! It's great to see benchmark tests.

**Review by mudit2812 — CHANGES_REQUESTED** [review]
Thanks for addressing all the comments so fast! I just left a few more comments pertaining to tests. I'm very happy to approve once these comments are addressed.

File: tests/ops/qubit/test_observables.py
Comment (andrijapau) [reply tests/ops/qubit/test_observables.py:336]: Agreed, good idea!

File: tests/ops/qubit/test_observables.py
Comment (mudit2812) [inline tests/ops/qubit/test_observables.py:369]: Sorry if it wasn't clear in my earlier review, I meant that you can test `Hermitian.decomposition` and `Hermitian.compute_decomposition` in the same test. This is not really a blocking request though.

```suggestion
        A_decomp_static = qml.Hermitian.compute_decomposition(observable, wires=list(range(num_wires)))
        A_decomp = qml.Hermitian(observable, wires=list(range(num_wires))).decomposition()

        assert np.allclose(A_decomp_static[0].matrix(), observable, rtol=0)
        assert np.allclose(A_decomp[0].matrix(), observable, rtol=0)
```

**Review by mudit2812 — APPROVED** [review]
Nice work @andrijapau ! Thanks for the fast responses and for adding this feature 🎉 

I'll request a 2nd review from the team.

File: tests/ops/qubit/test_observables.py
Comment (andrijapau) [reply tests/ops/qubit/test_observables.py:369]: I see! I'll go ahead and implement that and have it ready for the team to review 🙂.

**Review by albi3ro — APPROVED** [review]
Thanks for addressing @mudit2812 's comments 👍 The benchmarking graph looks great :) thanks for producing that.

No concerns from my side, and I'm happy to approve 🚀

**Comment by andrijapau** [comment]
Great! Looks like the PR is ready to be merged whenever you are.

Thank you all for your help and guidance and I'm excited to be a Pennylane contributor! 😊

**Comment by Alex-Preciado** [comment]
Thank you so much, @andrijapau! I appreciate the effort you put into the implementation. I have a few questions about your plot: I noticed you plotted the theoretical time complexity O(n 4^n). Could you explain why this is the expected complexity? Also, how should I interpret the horizontal bars on your plot?

**Comment by andrijapau** [comment]
Certainly, @Alex-Preciado!

> Could you explain why this is the expected complexity? 

The short answer is that I read it in the [documentation](https://docs.pennylane.ai/en/stable/code/api/pennylane.pauli_decompose.html). 

<details closed>
<summary>⏰ Open for Time Complexity Analysis ⏰</summary>
<br>
The long answer is that we can analyze the operations in `_generalized_pauli_decompose` that won't take constant time and determine the dominant time complexity.

```python
# Permute by XORing
indices = [qml.math.array(range(shape[0]))] 
for idx in range(shape[0] - 1): # O(2^n)
    indices.append(qml.math.bitwise_xor(indices[-1], (idx + 1) ^ (idx))) 

term_mat = qml.math.cast(
    qml.math.stack(
        [qml.math.gather(matrix[idx], indice) for idx, indice in enumerate(indices)]
    ),
    complex,
) # O((2^n)^2), stacking 2^n arrays of size 2^n.
```
Permuting indices would be overall $\mathcal{O}(4^n)$.

```python
# Perform Hadamard transformation on columns
hadamard_transform_mat = _walsh_hadamard_transform(qml.math.transpose(term_mat)) # O(n 2^n), from documentation

# Account for the phases from Y
phase_mat = qml.math.ones(shape, dtype=complex).reshape((2,) * (2 * num_qubits))
for idx in range(num_qubits):
    index = [slice(None)] * (2 * num_qubits)
    index[idx] = index[idx + num_qubits] = 1
    phase_mat[tuple(index)] *= 1j
phase_mat = qml.math.convert_like(qml.math.reshape(phase_mat, shape), matrix) # O(n)
```
WHT would be dominated by $\mathcal{O}(n 2^n)$.

```python
# c_00 + c_11 -> I; c_00 - c_11 -> Z; c_01 + c_10 -> X; 1j*(c_10 - c_01) -> Y
# https://quantumcomputing.stackexchange.com/a/31790
term_mat = qml.math.transpose(qml.math.multiply(hadamard_transform_mat, phase_mat)) # O((2^n)^2)
```
Element wise multiplication has complexity $\mathcal{O}(4^n)$.

```python
# Obtain the coefficients for each Pauli word
coeffs, obs = [], []
for pauli_rep in product("IXYZ", repeat=num_qubits):
    bit_array = qml.math.array(
        [[(rep in "YZ"), (rep in "XY")] for rep in pauli_rep], dtype=int
    ).T
    coefficient = term_mat[tuple(int("".join(map(str, x)), 2) for x in bit_array)]

    if not is_abstract(matrix) and qml.math.allclose(coefficient, 0):
        continue

    observables = (
        [(o, w) for w, o in zip(wire_order, pauli_rep) if o != I]
        if hide_identity and not all(t == I for t in pauli_rep)
        else [(o, w) for w, o in zip(wire_order, pauli_rep)]
    )
    if observables:
        coeffs.append(coefficient)
        obs.append(observables)

coeffs = qml.math.stack(coeffs)
```
I believe this is where the poor time complexity comes from. We need to (1) loop through all possible `pauli_rep` tensor products ($4^n$, four choices per qubit) and (2) create the `bit_array` by looping through the `pauli_rep` tuple of length $n$ 😱. With indexing being a constant time, the complexity of this section would be $\mathcal{O}(n 4^n)$. 

The total time complexity of this algorithm would be the sum of each of these operations $\mathcal{O}(4^n + n2^n + 4^n + n4^n)$ but would be dominated by $\mathcal{O}(n4^n)$ overall.

</details>

> Also, how should I interpret the horizontal bars on your plot?

The horizontal bars are auto-generated box-and-whisker plots for the performance of each hermitian decomposition test case.

When the benchmarking test is called, the following table is automatically generated,

benchmark-table.pdf

The reason there is uncertainty in the histogram bars is due to the `Rounds` metric. The benchmarking protocol will execute the test case multiple times to account for any variations in performance. As an illustration, I've outlined a test in white which took on average $\Delta t = 257\ \mu s$ to complete (statistics generated from the 826 rounds of execution).

**Comment by Alex-Preciado** [comment]
Awesome! Thank you so much for the detailed explanation, @andrijapau!! I’ll go ahead and update this branch and merge It 😎

**Pull request merged**
